numpy/_core/src/umath/umathmodule.c

```diff
@@ -33,31 +33,109 @@
 #include "special_integer_comparisons.h"
 #include "extobj.h"  /* for _extobject_contextvar exposure */
 #include "ufunc_type_resolution.h"
+#include "legacy_array_method.h"
 
 /* Automatically generated code to define all ufuncs: */
 #include "funcs.inc"
 #include "__umath_generated.c"
 
 
-static PyUFuncGenericFunction pyfunc_functions[] = {PyUFunc_On_Om};
+static int
+frompyfunc_promoter(PyObject *ufunc,
+        PyArray_DTypeMeta *const NPY_UNUSED(op_dtypes[]),
+        PyArray_DTypeMeta *const NPY_UNUSED(signature[]),
+        PyArray_DTypeMeta *new_op_dtypes[]);
 
 static int
-object_ufunc_type_resolver(PyUFuncObject *ufunc,
-                                NPY_CASTING casting,
-                                PyArrayObject **operands,
-                                PyObject *type_tup,
-                                PyArray_Descr **out_dtypes)
+frompyfunc_promoter(PyObject *ufunc,
+        PyArray_DTypeMeta *const NPY_UNUSED(op_dtypes[]),
+        PyArray_DTypeMeta *const NPY_UNUSED(signature[]),
+        PyArray_DTypeMeta *new_op_dtypes[])
 {
-    int i, nop = ufunc->nin + ufunc->nout;
+    int i, nop = ((PyUFuncObject *)ufunc)->nin + ((PyUFuncObject *)ufunc)->nout;
 
-    out_dtypes[0] = PyArray_DescrFromType(NPY_OBJECT);
-    if (out_dtypes[0] == NULL) {
+    new_op_dtypes[0] = PyArray_DTypeFromTypeNum(NPY_OBJECT);
+    if (new_op_dtypes[0] == NULL) {
         return -1;
     }
 
     for (i = 1; i < nop; ++i) {
-        Py_INCREF(out_dtypes[0]);
-        out_dtypes[i] = out_dtypes[0];
+        Py_INCREF(new_op_dtypes[0]);
+        new_op_dtypes[i] = new_op_dtypes[0];
+    }
+
+    return 0;
+}
+
+static int
+pyfunc_loop(PyArrayMethod_Context *context,
+                char *const data[], npy_intp const dimensions[],
+                npy_intp const strides[], NpyAuxData *NPY_UNUSED(auxdata))
```

Comment:
Too bad that we need to duplicate this, since the old version is publically exposed, but nothing to do about it :).